torch/csrc/autograd/engine.cpp

```diff
@@ -202,12 +214,20 @@ Engine::~Engine() {
   }
   if (noBackward) {
     for (auto& queue : ready_queues_) {
-     queue->pushShutdownTask();
+      for (int i = 0; i < num_threads_per_device_; ++i) {
+        queue->pushShutdownTask();
+      }
     }
   }
   // Othewise threads are leaked
 }
 
+int Engine::get_num_threads_per_device() { return num_threads_per_device_; }
+
+void Engine::set_num_threads_per_device(int num_threads_per_device) {
```

Comment:
(discussed offline) Current implementation only takes effect if setting #threads **before** the first run of autograd (which creates the threads). Better to add a check here to raise error if users attempt to set to a different #threads after the thread pool has been created.